pilot/pkg/model/context.go

```diff
@@ -101,26 +100,14 @@ type Proxy struct {
 
 	// Metadata key-value pairs extending the Node identifier
 	Metadata map[string]string
-
-	// mutex control access to mutable fields in the Proxy. On-demand will modify the
-	// list of services based on calls from envoy.
-	mutex sync.RWMutex
-
-	// serviceDependencies, if set, controls the list of outbound listeners and routes
```

Comment:
It would be really great to break large PRs - adding the Sidecar to galley and model is a clear change, 
deleting/changing/implementing can be in a separate PR so it's easy to review and merge.

Reply:
In particular - why is this removal needed ? It was used for isolation - which is still needed with Sidecar.

Reply:
Sorry, I mean this is intended for on-demand and isolation - each proxy will track the dynamic dependencies. 
On-demand is in progress - there is no need to delete its implementation.

Reply:
Sure - but no point to start with the bad data structures, you know how hard it is to change. 

They co-exist by having the dependencies in the proxy object, per node - instead of a static map as you define in this PR. 

We can probably use a static map as 'defaults' - but we do need the per-node.

Again - I am working on this and other people as well, you can't remove code and 
use an incompatible struct without a better reason.

Reply:
These are not the defaults. The sidecar API can be customized to individual workloads (the label based selection code is already in place and optimized). The code here is forming a materialized view of the registry, configs, etc. per listener in a sidecar object, taking into account the public/private scopes as well as the inheritance of destination rules *subject to imports*. Any proxy assigned to a sidecar object will see the right set of synthesized rules. 

The code you have here including the isolation stuff, is disregarding the import structure per listener as well as the rule inheritance. Simply having a list of services per proxy is insufficient - may be helpful for CDS but certainly not LDS or RDS. A simple example: I can have two public virtualServices in two namespaces, for the same service. The choice of imported namespace determines the right virtualService to use and in turn influences where the traffic will actually be routed as determined by the virtual service.

Reply:
With on-demand, each individual proxy will use a different set of services - based on the actual traffic. That's why the deps need to include proxy and the list of 'watched' resources that envoy manages. 

The 'isolation' code can read the Sidecar structs you create, and filter it based on what envoy 
is actually watching. 

On the second part: not sure how it is relevant. Yes, the implementation of Sidecar needs to import the right namespace. But the data structure can still be in the proxy so it can be combined with on-demand info. 

I'm not sure what is your point - that we should stop working on on-demand, or that it is possible 
to implement on-demand without keeping the deps in the proxy ?

Reply:
No. My point is ondemand and the proactive configuration using sidecar apis are two mutually exclusive things. It would be good to have a Doc that sketches and clearly lays out how the two models will interact. This is a big change impacting end user experience. If the proxy can be loaded on demand, there is no need for a sidecar config. So, lets discuss this structural change in a doc and sort it out first. You clearly have a vision in mind in terms of how the two should interact and work, and I am failing to see it. So I don't mind parking this PR until this is sorted out completely, as I feel that this whole sidecar api stuff is of no use when you can load listeners/routes/clusters on demand.